### PR TITLE
Marketplace logic part 2 (buying functionality)

### PR DESCRIPTION
**Pull request by lin8231**
* Can now buy. 
* If search term is invalid, will have pop up saying invalid search term
* If purchase amount invalid (less than 1 or greater than available amount) will have pop up saying invalid purchase amount
* If purchase goes through, sends request to backend with relevant data and resets the buying page and refetches the relevant info (which includes general info and investments info)
# New
* Refactored App.js data fetching. There is now also a partial fetch function. 
* When the user clicks buy and it goes through, the servlet data need to be **partially** refetched - in this case, only /investments and /competitionInfo. the partial fetch function fetches and updates the corresponding state variable for those two only. (it was refetching everything before)

### CODE REVIEW

File: project/frontend/src/containers/Competition/Competition.js [hunk 1]
```diff
@@ -3,120 +3,121 @@ import classes from "./Competition.module.css";
...
+  const toDayHourMinute = (totalTime) => {
+    let millisInDay = 24 * 60 * 60 * 1000,
+      millisInHour = 60 * 60 * 1000,
+      days = Math.floor(totalTime / millisInDay),
+      hours = Math.floor((totalTime - days * millisInDay) / millisInHour),
+      minutes = Math.round(
+        (totalTime - days * millisInDay - hours * millisInHour) / 60000
+      ),
+      pad = function (n) {
+        return n < 10 ? "0" + n : n;
```
Comment (roonil-wazlib) [inline project/frontend/src/containers/Competition/Competition.js:18]: what in god's name does this do

**Review by roonil-wazlib — APPROVED** [review]
Apart from the fact that I hate ternary statements, I approve.

File: project/frontend/src/containers/Competition/Competition.js (hunk 1 above)
Comment (lin8231) [reply project/frontend/src/containers/Competition/Competition.js:18]: soo if it's 11:05 (as in 11 hours and 5 minutes) it pads 5 minutes to 05 minutes

Comment (lin8231) [reply project/frontend/src/containers/Competition/Competition.js:18]: wait isn't this your code? XD I didn't touch that file at all, it's just my formatter changing the formatting no code changes

**Review by texm — APPROVED** [review]
LGTM, a few debuggy things lying around but we can clean that up later.

File: project/frontend/src/containers/Competition/Competition.js [hunk 2]
```diff
@@ -3,120 +3,121 @@ import classes from "./Competition.module.css";
...
+    return [days, pad(hours), pad(minutes)];
+  };
 
-    const getTimeRemaining = () => {
-        const millisNow = Date.now();
-        const remainingTime = competitionEndDate - millisNow;
-        return formatDHM(toDayHourMinute(remainingTime));
-    };
+  const formatDHM = (dhm) => {
+    return dhm[0] + "  Days  " + dhm[1] + "  Hours  " + dhm[2] + "  Minutes  ";
```
Comment (texm) [inline project/frontend/src/containers/Competition/Competition.js:32]: might look nicer as a template string

**Pull request merged**